### ISSUE TITLE
Extend reduction from forecasting to regression to handle exogenous data 

### ISSUE THREAD

**Issue by mloning** [alan-turing-institute/sktime#345]
**Is your feature request related to a problem? Please describe.**
Currently, the reduction meta-estimators only handle univariate data (we raise a not-implemented error [here](https://github.com/alan-turing-institute/sktime/blob/979aed585983fb820ddb0ca8a619d5730dbed4a7/sktime/forecasting/compose/_reduce.py#L67)). We need to extend it to support multivariate time series data. 

**Describe the solution you'd like**
* Use temporal cross-validation iterator to transform multivariate time series into data frame with lagged values and vector with target values. Currently, the transformation happens [here](https://github.com/alan-turing-institute/sktime/blob/979aed585983fb820ddb0ca8a619d5730dbed4a7/sktime/forecasting/compose/_reduce.py#L73). 
* The `x_windows` list needs to be extended to contain the data from all series. 
* The different formatting of the windows for reduction to time series and tabular regression needs to be adapted too to handle multivariate data. 

**Additional context**
Linked to #228

**Comment by btrtts** [comment]
take

**Comment by mloning** [comment]
The code is here: https://github.com/alan-turing-institute/sktime/blob/master/sktime/forecasting/compose/_reduce.py

The current code isn't great because:
* convoluted class hierarchy makes it hard to follow and extend
* inefficient for-loops for data transformations 

I suggest we break up the code into simpler classes using stand-alone functions for the data transformations. 

@SebasKoel has done some work already here: https://github.com/alan-turing-institute/sktime/pull/390 (see e.g. `create_lag_matrix`) 

I suggest we start with the reduction to the simpler tabular regression setting (basically scikit-learn, i.e. transforming time series into 2d numpy array) and work on the time series regression setting (3d numpy) once the tabular setting is working. 

What we want to be able to do in the end is something like this: 

```
from sktime.forecasting.all import * 
from sklearn.ensemble import RandomForestRegressor

y, X = load_longley()
y_train, y_test, X_train, X_test = temporal_train_test_split(y, X)
fh = ForecastingHorizon(y_test.index, is_relative=False)

regressor = RandomForestRegressor()
forecaster = ReducedRegressionForecaster(regressor, window_length=10, strategy="recursive")
forecaster.fit(y_train, X_train)
y_pred = forecaster.predict(fh, X=X_test)
```

Note that we not only create lagged values for the target series (`y`) but also the exogenous data (`X`). We then pass all lagged values in the required 2d numpy array to scikit-learn regressor.

**Comment by bradyneal** [comment]
Any chance there's an ETA on this? This seems like a pretty essential feature for people deciding which forecasting library to use.

**Comment by mloning** [comment]
Hi @bradyneal, yes I agree! Do you want to take a look at this? What other libraries are you considering?

**Comment by bradyneal** [comment]
Maybe in the future (lots of uncertainty), but not right now, unfortunately. PyTorch Forecasting, GluonTS, and NeuralProphet. From a preliminary look, `sktime` seems the best in terms of how well designed it is, its broad scope, and lots of great classical forecasters :), but not in other aspects such as exogenous data support (beyond AutoARIMA), more advanced hyperparameter tuning, and SOTA neural models (not in sktime-dl).

**Comment by mloning** [comment]
@bradyneal yeah, I agree, still early days for sktime and lots of work to be done.

**Comment by bradyneal** [comment]
Simple integration with a more efficient hyperparameter tuning library such as Optuna (e.g. PyTorch Forecasting does this). Random search is nice to have over grid search, but I don't think either of them really compare to more modern hyperparameter tuning algorithms. However, I don't think sklearn even has this integration.

**Comment by mloning** [comment]
Interesting, wasn't aware of Optuna, will take a look!

**Comment by mloning** [comment]
Perhaps we should make the reduction compositor more pipeline like, allowing users to pass transformers to specify how to transform the data. This way we could also handle different transformations for different (exogenous) variables. 

Something like this:  
```python
forecaster = ReducedRegressionForecaster(regressor, transformers)
```
Together with series-to-tabular transformer type for the sliding window stacking transformation.

**Comment by fkiraly** [comment]
Hm, why would you want to do this instead of `ReducedRegressionForecaster(Pipeline(stuff))` where `Pipeline` is directly from `sklearn`?

**Comment by mloning** [comment]
@fkiraly that wouldn't work, because you need the sliding-window transformation to go to the tabular format required for the sklearn pipeline. The regressor itself could of course be a sklearn pipeline.

**Comment by mloning** [comment]
@Lovkush-A here's some code how we could handle exogenous variables in the sliding-window transformation, it also speeds up the transformation. 

```python
import numpy as np
from sktime.forecasting.model_selection import temporal_train_test_split
from sktime.datasets import load_longley
y, X = load_longley()
y = y.astype(np.float)
X = X.astype(np.float)
X.index = y.index
y_train, y_test, X_train, X_test = temporal_train_test_split(y, X)
print(y_train.shape, y_test.shape, X_train.shape, X_test.shape)
from sktime.forecasting.all import *
from sklearn.ensemble import RandomForestRegressor
reducer = ReducedRegressionForecaster(RandomForestRegressor(), window_length=3)
window_length = 10
series = pd.Series(np.arange(10_000), dtype=np.float)
# our current approach
reducer._cv = SlidingWindowSplitter(start_with_window=True, window_length=window_length)
%timeit reducer._transform(series)
Xa, ya = reducer._transform(series)
Xa = Xa[:, ::-1]
# statsmodels functionality
from statsmodels.tsa.tsatools import lagmat, lagmat2ds
def create_lag_matrix_statsmodels(y, X=None, window_length=10, **kwargs):
    if X is None:
        ret = lagmat(y, window_length, trim="backward", use_pandas=False, **kwargs)
    else:
        ret = lagmat2ds(pd.concat([y, X], axis=1), window_length, trim="backward", use_pandas=False, **kwargs)
    ret = ret[:-window_length + 1, :]
    return ret[:, 0].reshape(-1, 1), ret[:, 1:]
%timeit create_lag_matrix_statsmodels(series, window_length=window_length + 1)
yb, Xb = create_lag_matrix_statsmodels(series, window_length=window_length + 1)
np.testing.assert_array_equal(ya, yb)
np.testing.assert_array_equal(Xa, Xb)
# revision of our approach
def create_lag_matrix(x, window_length):
    # input checks
    x = np.asarray(x)
    if x.ndim == 1:
        x = x.reshape(-1, 1)
    # time series tabularisation
    n_timepoints, n_variables = x.shape
    lm = np.zeros((n_timepoints + window_length, n_variables * (window_length + 1)))
    for k in range(window_length + 1):
        a = window_length - k
        b = n_timepoints + window_length - k
        c = n_variables * (window_length - k)
        d = n_variables * (window_length - k + 1)
        lm[a:b, c:d] = x
    # truncate
    lm = lm[window_length:-window_length]
    # separate y, X
    y = lm[:, 0].reshape(-1, 1)
    X = lm[:, 1:]
    return y, X
%timeit create_lag_matrix(series, window_length)
yc, Xc = create_lag_matrix(series, window_length)
np.testing.assert_array_equal(yc, ya)
np.testing.assert_array_equal(Xc, Xa)
```
